### DISCUSSION TITLE
Mixcr to VDJtools

### DISCUSSION THREAD

**Discussion by pbpayal** [milaboratory/mixcr#1617] [Q&A] [answered]
## takara-mouse-rna-bcr-smarter
I used this module to generate results from Mixcr. 

## Exact MiXCR commands 

``mixcr analyze takara-mouse-rna-bcr-smarter 1B_fastq_R1.gz  1B_fastq_R2.gz result_1B``

### Import into VDJtools
I want to further investigate the results using VDJtools. 
*Which output file to use for vdjtools? Is it the .clns file or .tsv files? 
*Do I need to convert the file for VDJtool usage? 
*I tried both convertion and using the .tsv file as is in VDJtools. It gives me error.

### Actual Result Files
result_1B.align.report.basicstats.txt
result_1B.align.report.json
result_1B.align.report.txt
result_1B.assemble.report.json
result_1B.assemble.report.txt
result_1B.clns
result_1B.clones_IGH_nondecimal.tsv
result_1B.clones_IGH_nondecimal.txt
result_1B.clones_IGH.tsv
result_1B.clones_IGK.tsv
result_1B.clones_IGL.tsv
result_1B.qc.json
result_1B.qc.txt
result_1B.vdjca

### VDJ command Convert for TSV file
*````
java -Xmx4g -jar $VDJ_JAR Convert -S mixcr result_1B.clones_IGH.tsv result_1B.clones_IGH ````*

### Head of output TSV file (IGH)

*````cloneId	readCount	readFraction	targetSequences	targetQualities	allVHitsWithScore	allDHitsWithScore	allJHitsWithScore	allCHitsWithScore	allVAlignments	allDAlignments	allJAlignments	allCAlignments	nSeqFR1	minQualFR1	nSeqCDR1	minQualCDR1	nSeqFR2	minQualFR2	nSeqCDR2	minQualCDR2	nSeqFR3	minQualFR3	nSeqCDR3	minQualCDR3	nSeqFR4	minQualFR4	aaSeqFR1	aaSeqCDR1	aaSeqFR2	aaSeqCDR2	aaSeqFR3	aaSeqCDR3	aaSeqFR4	refPoints
1	207.0	0.350253807106599	GTTGTAATTAGTATCACCATCTCCAGGAAAAATCTGTCCAATCCACTCAAGACCCTTTCCAGGCCTCTGCTTCACCCAGTTCATCCAGTAATTACTGGATGAACTGGGTGAAGCAGAGGCCTGGAAAGGGTCTTGAGTGGATTGGACAGATTTTTCCTGGAGATGGTGATACTAATTACAACGGAAAATTCAAGGACAAGGCCACACTGACTGCAGACAAATCCTCCAACACAGCCTACATGCAGCTCAGCAGTCTGACCTCTGAGGACTCTGCGGTCTATTTCTGTGCAAGAGGCGGCTACTGGGGCCAAGGCACCACTCTCACAGTCTCCTCAG	[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[	IGHV1-80*00(676)		IGHJ2*00(380)	IGHG3*00(181.7)	77|371|391|0|293|SC77GSA78TSG79TI82AI82ASC83TSC86TST87ASG88TSG91CSG94TI95CST97CSG98ASG101ASC102AST103ASG104ASG106TSG110TST111CSG112CSG115TST118ADG119DG120DG121DG122DC123SG128ADT129SA132CSG133CSA134CST140AI142GI142CSA143TSA144CSA145TST151ASG152CSG153CST155ASA156GSC157TSG158TST162CSG168ASC169TSA231TSC253TSG265ASG273ASG306ASC331T|1199.0		30|68|68|298|336||380.0		GTTGTAATTAGTATCACCATCTCCAGGAAAAATCTGTCCAATCCACTCAAGACCCTTTCCAGGCCTCTGCTTCA	58	CCCAGTTCATCCAGTAATTACTGG	58	ATGAACTGGGTGAAGCAGAGGCCTGGAAAGGGTCTTGAGTGGATTGGACAG	58	ATTTTTCCTGGAGATGGTGATACT	58	AATTACAACGGAAAATTCAAGGACAAGGCCACACTGACTGCAGACAAATCCTCCAACACAGCCTACATGCAGCTCAGCAGTCTGACCTCTGAGGACTCTGCGGTCTATTTC	58	TGTGCAAGAGGCGGCTACTGG	58	GGCCAAGGCACCACTCTCACAGTCTCCTCAG	58	VVISITISRKNL_QSTQDPFQASAS	PSSSSNYW	MNWVKQRPGKGLEWIGQ	IFPGDGDT	NYNGKFKDKATLTADKSSNTAYMQLSSLTSEDSAVYF	CARGGYW	GQGTTLTVSS_	::::0:74:98:149:173:284:0:293:::::298:-10:305:336::````*

### Error
*````
[ERROR] java.lang.RuntimeException: Unable to parse clonotype string 207 for MiXcr input type: Some mandatory columns are absent in the input file.````  *

**Comment by mizraelson** [comment, marked as answer]
Hi,
VDJtools is our legacy software, so we don't support it anymore. The new MiXCR has a slightly new output format that's why `vdjtools` does not recognize it. Most of the [postanalysis](https://mixcr.com/mixcr/reference/mixcr-postanalysis/) function are now included in the MiXCR so you might wanna try using it instead.

Nevertheless, there is a way to manually bypass the issue if needed:
- New MiXCR has columns named : `readCount`, `readFraction` and `uniqueMoleculeCount`, `uniqueMoleculeFraction` (in case of UMIs).  For `vdjtools` to work correctly the following columns are required: `cloneCount`, `cloneFraction`. A simple rename will help. Keep in mind that if you have both reads and UMIs you should pick which one do you wanna use and rename the columns of interest.
- Another thing is `region not covered` field that sometimes is present if gene region was not covered by the alignment. We used to leave these fields blank (if the region was not covered) but that confused some of our users so we added this note.
As a bypass you can run the following command for the `mixcr` tsv output files to remove it:
    `sed -i 's/region_not_covered//g' sample1.tsv`

**Reply by pbpayal** [reply]
Thanks, I will give a shot at postanalysis too.

**Comment by pbpayal** [comment]
Hello,

I have changed the column names. I am getting a different error.

`Unknown` symbol "0", see _vdjtools_error.log for details

I tried removing .0 from columns readCount, allVAlignments, allDAlignments, allUAlignments

How to solve this?

Thanks

**Reply by mizraelson** [reply]
Did you open the table in Excel to rename the columns? Sometimes, doing that leads to formatting errors that Excel introduces. If that is the case, I would suggest using a regular text editor to rename the columns there. 
If the error persists, could you share this table with me so I can take a look?

**Reply by pbpayal** [reply]
Hi, 

I changed the column names using vi editor as well as tried sed. I still get some other error. 

`Unable to parse clonotype string 3057	19.0	1.0011697878573913E-4	CAGATCCAGTTGGTGCAGTCTGGACCTGAGCTGAAGAAGCCTGGAGAGACAGTCAAGATCTCCTGCAAGGCTTCTGGGTATACCTTCACAGAATATCCAATGCACTGGGTGAAGCGGGCTCCAGGAAAGGGTTTCAAGTGGATGGGCATGATATACACCGACACTGGAGAGCCAACATATGCTGAAGAGTTCAAGGGACGGTTTGCCTTCTCTTTGGAGACCTCTGCCAGCACTGCCTATTTGCAGATCAACAACCTCAAAAATGAGGACACGGCTACATATTTTTTATGGTACCGTGACTACTGGGGCCAAGGCACCACTCTCACAGTCTCCTCAG	[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[	IGHV9-1*00(3525)	IGHD2-8*00(35),IGHD2-7*00(30)	IGHJ2*00(410)	IGHG3*00(190.3)	77|361|391|0|284|SA192G|2811.0	22|29|51|286|293||35.0;22|28|51|286|292||30.0	27|68|68|296|337||410.0		CAGATCCAGTTGGTGCAGTCTGGACCTGAGCTGAAGAAGCCTGGAGAGACAGTCAAGATCTCCTGCAAGGCTTC58	GGGTATACCTTCACAGAATATCCA	58	ATGCACTGGGTGAAGCGGGCTCCAGGAAAGGGTTTCAAGTGGATGGGCATG	58	ATATACACCGACACTGGAGAGCCA	58	region_not_covered	region_not_covered	region_not_covered	region_not_covered	GGCCAAGGCACCACTCTCACAGTCTCCTCAG	58	QIQLVQSGPELKKPGETVKISCKAS	GYTFTEYP	MHWVKRAPGKGFKWMGM	IYTDTGEP	region_not_covered	region_not_covered	GQGTTLTVSS_	::::0:75:99:150:174::-10:284:286:-5:-5:293:296:-7:306:337:: for MiXcr input type: Unknown symbol "e", see _vdjtools_error.log for details`


Attached my clone file (after renaming columns) and error file.

Thanks,
Payal


[vdjtools_error.zip](https://github.com/milaboratory/mixcr/files/15490280/vdjtools_error.zip)

**Reply by mizraelson** [reply]
Yeah, you see the values like `1.0011697878573913E-4`. The original file doesn't have it. This appears when you open and sav the file in the editor.

**Reply by pbpayal** [reply]
I actually just opened the file using Apple numbers to check the file headers before uploading the file...i didn't use excel to edit the header. Let me check again.

**Reply by mizraelson** [reply]
The original file generated by VDJtools doesn't have `E-4` part, instead there will be a number like: `0.0000001169...`. And that is what causing the error.

**Reply by pbpayal** [reply]
Please check this attached file. I used vi editor to change the column names. 
[result_1B.clones_IGH.tsv.zip]
(https://github.com/milaboratory/mixcr/files/15490417/result_1B.clones_IGH.tsv.zip)
Do you still see E-4 numbers in this file?

 I still get the error.

**Reply by mizraelson** [reply]
No, but I see `region_not_covered`. As mentioned above you should remove this leaving an empty line, for example using `sed -i 's/region_not_covered//g' sample1.tsv` command.

**Reply by pbpayal** [reply]
Thank you. That works.

**Reply by jialin-li-cell** [reply]
ljl@Jerry:~/vdjtools-1.2.1$ java --illegal-access=deny --add-opens java.base/java.lang=ALL-UNNAMED -jar vdjtools-1.2.1.jar Convert -S mixcr -m /home/ljl/TRBnew/TRB.immmetadata.txt  VDJtools
Executing com.antigenomics.vdjtools.misc.Convert -S mixcr -m /home/ljl/TRBnew/TRB.immmetadata.txt VDJtools
[Sat Aug 17 20:12:54 CST 2024 Convert] Reading sample(s)
[Sat Aug 17 20:12:54 CST 2024 Convert] 16 sample(s) loaded
[Sat Aug 17 20:12:54 CST 2024 SampleStreamConnection] Loading sample 1
[ERROR] java.lang.RuntimeException: Unable to parse clonotype string 0  2391.0  4.2617487286495927E-4   TGTGCCAGCAGCTTAGCGGGCACAGATACGCAGTATTTT [[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[[ TRBV7-9*00(369.6)               TRBJ2-3*00(477.8)      476|494|513|0|18||180.0          21|41|69|19|39||200.0           TGTGCCAGCAGCTTAGCGGGCACAGATACGCAGTATTTT 58      CASSLAGTDTQYF   :::::::::0:1:18:::::19:-1:39::: for MiXcr input type: Some mandatory columns are absent in the input file., see _vdjtools_error.log for details

i dont konw how to change 4.2617487286495927E-4  into 0.00000 in ubuntu lunix system , for i cant use excel but open with text editor, using could u please tell me some code used in terminal to acheive that?  thanks!!!!!

**Reply by jialin-li-cell** [reply]
i solve this,thank god!

**Reply by linqy-immune** [reply]
hello!!! I encountered the ERROR "[ERROR] java.lang.RuntimeException: Unable to parse clonotype string 0" like yours, could you please tell me how you solved this!!! Thank you very much!!!

**Reply by mizraelson** [reply]
Hi,
Please refer to the original answer in this thread. You will need to rename the columns and remove the `region not covered` values.

**Reply by linqy-immune** [reply]
Thanks. I've already tried the tips you mentioned above, but it didn't take effect. Could you please help me have a look at my file and error_log. Thank you very much!!!
[_vdjtools_error.log](https://github.com/user-attachments/files/18394127/_vdjtools_error.log)

**Reply by mizraelson** [reply]
Do you have `cloneCount` and `cloneFraction` columns in your file?

**Reply by linqy-immune** [reply]
I've figured out this! Thank you very much!!! Best  wishes!

**Reply by linqy-immune** [reply]
@mizraelson 
Hi,
I'm soooo sorry to interrupt you again, and really grateful to use such an excellent and great software although it has been a legacy sofaware. I have no other choices as I can't handle this for a long time. Could you please give me some suggestions?
There are a group of files downloaded from AdaptiveBiotech, and the column names are"rearrangement extended_rearrangement bio_identity amino_acid templates frame_type rearrangement_type productive_frequency cdr1_start_index cdr1_rearrangement_length cdr2_start_index cdr2_rearrangement_length cdr3_start_index cdr3_length v_index n1_index d_index n2_index j_index v_deletions n2_insertions d3_deletions d5_deletions n1_insertions j_deletions chosen_j_allele chosen_j_family chosen_j_gene chosen_v_allele chosen_v_family chosen_v_gene d_allele d_allele_ties d_family d_family_ties d_gene d_gene_ties d_resolved j_allele j_allele_ties j_family j_family_ties j_gene j_gene_ties j_resolved v_allele v_allele_ties v_family v_family_ties v_gene v_gene_ties v_resolved" and I tried to analyze these data by vdjtools but it seems that the data format don't match vdjtools' running demand. The code I run was `java -jar D:\vdjtools\vdjtools.jar Convert -S immunoseq "D:\COV9\P0008111_TCRB.tsv" output_prefix` and the ERROR was `java.lang.RuntimeException: Unable to parse clonotype string CTGGAGTCTGCTGCCTCCTCCCAGACATCTGTATATTTCTGCGCCAGCAGTGAGGCTCCGGGGGCGGGAGAGCAGTACTTCGGGCCG    ATGGGCACGAGGCTCTTCTTCTATGTGGCCCTTTGTCTGCTGTGGGCAGGACACAGGGATGCTGAAATCACCCAGAGCCCAAGACACAAGATCACAGAGACAGGAAGGCAGGTGACCTTGGCGTGTCACCAGACTTGGAACCACAACAATATGTTCTGGTATCGACAAGACCTGGGACATGGGCTGAGGCTGATCCATTACTCATATGGTGTTCAAGACACTAACAAAGGAGAAGTCTCAGATGGCTACAGTGTCTCTAGATCAAACACAGAGGACCTCCCCCTCACTCTGGAGTCTGCTGCCTCCTCCCAGACATCTGTATATTTCTGCGCCAGCAGTGAGGCTCCGGGGGCGGGAGAGCAGTACTTCGGGCCGGGCACCAGGCTCACGGTCACAG   CASSEAPGAGEQYF+TCRBV10-01+TCRBJ02-07   CASSEAPGAGEQYF   16      In      VDJ     2.1762489628813537E-4   135     15      201     18      327     42      39     54       59      65      69      2       5       no data 6       4       7       01      TCRBJ02 07      01      TCRBV1001       unknown no data unknown TCRBD01,TCRBD02 unresolved      TCRBD01-01,TCRBD02-01   unknown 01      no data TCRBJ02no data  TCRBJ02-07      no data TCRBJ02-07*01   01      no data TCRBV10 no data TCRBV10-01      no data TCRBV10-01*01 for ImmunoSeq input type: Some mandatory columns are absent in the input file., see _vdjtools_error.log for details`. I have attempted to resolve this with many methods but got nothing.
I really want to analyze the data by vdjtools because it is such an extraordinary software with many functions I need. I would appreciate it if you offer me some help. I am looking forward to your reply.
Thanks and best regards,
Linqy

**Reply by mizraelson** [reply]
Since ImmunoSeq is not our software, I’m not sure how the format has changed since we implemented this. You can find information about the required formatting and access a sample here:

https://vdjtools-doc.readthedocs.io/en/master/input.html#supported-input

**Reply by linqy-immune** [reply]
Thank you for your help! It is very kind of you, I would have a look at this.😊
Best wishes!
